debian/fop.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.14.
.TH FOP "1" "July 2020" "fop Version 2.5" "User Commands"
.SH NAME

fop \- Command line wrapper for the java version of fop.
.SH DESCRIPTION
Apache FOP (Formatting Objects Processor) is a print formatter driven by XSL
formatting objects (XSL-FO) and an output independent formatter. It is a Java
application that reads a formatting object (FO) tree and renders the resulting
pages to a specified output. Output formats currently supported include PDF,
PS, PCL, AFP, XML (area tree representation), Print, AWT and PNG, and to a
lesser extent, RTF and TXT. The primary output target is PDF.
.PP
.SH SYNOPSIS

fop [options] [\-fo|\-xml] infile [\-xsl file] [\-awt|\-pdf|\-mif|\-rtf|\-tiff|\-png|\-pcl|\-ps|\-txt|\-at [mime]|\-print] <outfile>
.IP
.SH OPTIONS
.TP
\fB\-version\fR
print FOP version and exit
.TP
\fB\-x\fR
dump configuration settings
.TP
\fB\-c\fR cfg.xml
use additional configuration file cfg.xml
.TP
\fB\-l\fR lang
the language to use for user information
.TP
\fB\-nocs\fR
disable complex script features
.TP
\fB\-r\fR
relaxed/less strict validation (where available)
.TP
\fB\-dpi\fR xxx
target resolution in dots per inch (dpi) where xxx is a number
.TP
\fB\-s\fR
for area tree XML, down to block areas only
.TP
\fB\-v\fR
run in verbose mode (currently simply print FOP version and continue)
.TP
\fB\-o\fR [password]
PDF file will be encrypted with option owner password
.TP
\fB\-u\fR [password]
PDF file will be encrypted with option user password
.TP
\fB\-noprint\fR
PDF file will be encrypted without printing permission
.TP
\fB\-nocopy\fR
PDF file will be encrypted without copy content permission
.TP
\fB\-noedit\fR
PDF file will be encrypted without edit content permission
.TP
\fB\-noannotations\fR
PDF file will be encrypted without edit annotation permission
.TP
\fB\-nofillinforms\fR
PDF file will be encrypted without fill in interactive form fields permission
.TP
\fB\-noaccesscontent\fR
PDF file will be encrypted without extract text and graphics permission
.TP
\fB\-noassembledoc\fR
PDF file will be encrypted without assemble the document permission
.TP
\fB\-noprinthq\fR
PDF file will be encrypted without print high quality permission
.TP
\fB\-a\fR
enables accessibility features (Tagged PDF etc., default off)
.TP
\fB\-pdfprofile\fR prof
PDF file will be generated with the specified profile
(Examples for prof: PDF/A\-1b or PDF/X\-3:2003)
.TP
\fB\-conserve\fR
enable memory\-conservation policy (trades memory\-consumption for disk I/O)
(Note: currently only influences whether the area tree is serialized.)
.TP
\fB\-cache\fR
specifies a file/directory path location for the font cache file
.TP
\fB\-flush\fR
flushes the current font cache file
.IP
.SH INPUT
.TP
infile
xsl:fo input file (the same as the next)
(use '\-' for infile to pipe input from stdin)
.TP
\fB\-fo\fR
infile       xsl:fo input file
.TP
\fB\-xml\fR infile
xml input file, must be used together with \fB\-xsl\fR
.TP
\fB\-atin\fR infile
area tree input file
.TP
\fB\-ifin\fR infile
intermediate format input file
.TP
\fB\-imagein\fR infile
image input file (piping through stdin not supported)
.TP
\fB\-xsl\fR stylesheet
xslt stylesheet
.TP
\fB\-param\fR name value <value> to use for parameter <name> in xslt stylesheet
(repeat '\-param name value' for each parameter)
.TP
\fB\-catalog\fR
use catalog resolver for input XML and XSLT files
.IP
.SH OUTPUT
.TP
outfile
input will be rendered as PDF into outfile
(use '\-' for outfile to pipe output to stdout)
.TP
\fB\-pdf\fR outfile
input will be rendered as PDF (outfile req'd)
.TP
\fB\-pdfa1b\fR outfile
input will be rendered as PDF/A\-1b compliant PDF
(outfile req'd, same as "\-pdf outfile \fB\-pdfprofile\fR PDF/A\-1b")
.TP
\fB\-awt\fR
input will be displayed on screen
.TP
\fB\-rtf\fR outfile
input will be rendered as RTF (outfile req'd)
.TP
\fB\-pcl\fR outfile
input will be rendered as PCL (outfile req'd)
.TP
\fB\-ps\fR outfile
input will be rendered as PostScript (outfile req'd)
.TP
\fB\-afp\fR outfile
input will be rendered as AFP (outfile req'd)
.TP
\fB\-tiff\fR outfile
input will be rendered as TIFF (outfile req'd)
.TP
\fB\-png\fR outfile
input will be rendered as PNG (outfile req'd)
.TP
\fB\-txt\fR outfile
input will be rendered as plain text (outfile req'd)
.TP
\fB\-at\fR [mime] out
representation of area tree as XML (outfile req'd)
specify optional mime output to allow the AT to be converted
to final format later
.TP
\fB\-if\fR [mime] out
representation of document in intermediate format XML (outfile req'd)
specify optional mime output to allow the IF to be converted
to final format later
.TP
\fB\-print\fR
input file will be rendered and sent to the printer
see options with "\-print help"
.TP
\fB\-out\fR mime outfile input will be rendered using the given MIME type
(outfile req'd) Example: "\-out application/pdf D:\eout.pdf"
(Tip: "\-out list" prints the list of supported MIME types and exits)
.TP
\fB\-svg\fR outfile
input will be rendered as an SVG slides file (outfile req'd)
Experimental feature \- requires additional fop\-sandbox.jar.
.TP
\fB\-foout\fR outfile
input will only be XSL transformed. The intermediate
XSL\-FO file is saved and no rendering is performed.
(Only available if you use \fB\-xml\fR and \fB\-xsl\fR parameters)
.IP
.SH EXAMPLES
.IP

fop foo.fo foo.pdf

fop \fB\-fo\fR foo.fo \fB\-pdf\fR foo.pdf (does the same as the previous line)

fop \fB\-xml\fR foo.xml \fB\-xsl\fR foo.xsl \fB\-pdf\fR foo.pdf

fop \fB\-xml\fR foo.xml \fB\-xsl\fR foo.xsl \fB\-foout\fR foo.fo

fop \fB\-xml\fR \- \fB\-xsl\fR foo.xsl \fB\-pdf\fR \-

fop foo.fo \fB\-mif\fR foo.mif

fop foo.fo \fB\-rtf\fR foo.rtf

fop foo.fo \fB\-print\fR

fop foo.fo \fB\-awt\fR
.SH HYPHENATION

fop can make use of external hyphenation rules that can be supplied in
the form of a \fBhyphenation jar\fR. You can find such a one at
\fIhttp://offo.sourceforge.net/\fR.

To actually use that, set the \fBFOP_HYPHENATION_PATH\fR environment variable
to the full path of this file, or use configuration files, see below.
.SH "CONFIGURATION FILES"

fop reads two configuration files on startup: \fI/etc/fop.conf\fR and
\fI$HOME/.foprc\fR. These are plain shell script files where you can define
for instance the \fBFOP_HYPHENATION_PATH\fR variable.

Another way to extend is \fIconf.d/\fR files. Simply add a configuration file
into \fI/etc/fop.conf.d/my.conf\fR.
.SH "JAVA-WRAPPERS NOTE"
This program is a shell script wrapper based on \fBjava-wrappers(7)\fR. You
therefore benefit from several features; please see the \fBjava-wrappers(7)\fR
manual page for more information about them.
.SH AUTHOR
Wrapper author: Ola Lundqvist <opal@debian.org>

This manual page was written by Mathieu Malaterre <malat@debian.org>,
for the Debian project (and may be used by others).
.SH "SEE ALSO"
.BR fop-ttfreader(1),
.BR java-wrappers(7).


debian/changelog
fop (1:2.5-3) unstable; urgency=medium

  * Team upload
  * Update SOURCE_DATE_EPOCH patch (Closes: #978499)
    - Conditionally use SOURCE_DATA_EPOCH in PDFInfo, PDFMetadata,
      PDFRenderingUtil, and FileIDGenerator classes.
    - Add try/catch logic for unparsable SOURCE_DATE_EPOCH

 -- tony mancill <tmancill@debian.org>  Thu, 31 Dec 2020 11:24:15 -0800

fop (1:2.5-2) unstable; urgency=medium

  * Team upload
  * Use SOURCE_DATE_EPOCH from system environment (Closes: #978499)
    For reproducible builds of packages that use fop as a build-dep.
  * Remove deprecated get-orig-source target
  * Update debian/watch to version 4
  * Bump Standards-Version to 4.5.1
  * Set Rules-Requires-Root: no in debian/control
  * Use debhelper-compat 13

 -- tony mancill <tmancill@debian.org>  Mon, 28 Dec 2020 11:42:20 -0800

fop (1:2.5-1) unstable; urgency=medium

  * d/maven.rules: Re-add missing qdox package
  * d/fop.1: Fix typo in manpage
  * New upstream version 2.5
  * d/patches: Refresh patches
  * d/fop.1: Update man page to latest
  * d/rules: Adapt to new 2.5 release

 -- Mathieu Malaterre <malat@debian.org>  Wed, 01 Jul 2020 13:44:32 +0000

fop (1:2.4-3) unstable; urgency=medium

  * d/rules: Properly copy mvn output files to ant expected location. Closes: #932734
  * d/rules: Minor cleanup when building javadoc
  * d/control: Remove reference to avalon-framework. Closes: #873209

 -- Mathieu Malaterre <malat@debian.org>  Thu, 05 Mar 2020 17:42:07 +0100

fop (1:2.4-2) unstable; urgency=medium

  * d/patches: Add patch to support JDK 11

 -- Mathieu Malaterre <malat@debian.org>  Mon, 17 Feb 2020 19:12:05 +0100

fop (1:2.4-1) unstable; urgency=medium

  [ tony mancill ]
  * Depend on fontbox2 instead of fontbox (Closes: #584472)
    Thank you to Philipp Hahn for the patch and update to the bug report.

  [ Mathieu Malaterre ]
  * New upstream version 2.4.
  * d/control: Make sure to use latest batik

 -- Mathieu Malaterre <malat@debian.org>  Sun, 16 Feb 2020 21:21:39 +0100

fop (1:2.3-2) unstable; urgency=medium

  [ Mathieu Malaterre ]
  * Install fop.xsd in the binary package. Closes: #847443

  [ Vincent Fourmond ]
  * Remove me from the uploaders. Closes: #897287

 -- Mathieu Malaterre <malat@debian.org>  Wed, 28 Nov 2018 09:59:21 +0100

fop (1:2.3-1) experimental; urgency=medium

  [ Mathieu Malaterre ]
  * New upstream version 2.3. Closes: #858816
  * Inject the output of mh_make
  * Update to debhelper 11
  * Update to use maven depends mechanism
  * Incorporate as dh original d/rules
  * Remove ant associated file
  * remove qdox for now
  * Remove deprecated instruction about pom file
  * Update location now that fop is in subfolder
  * Add missing artifacts required for testing
  * Make sure to use debian version for mockito
  * ignore test failure until investigation by upstream
  * refresh patches to cope with new directory structure
  * remove patch applied upstream
  * Explicit reference to debian/README.Debian in fop package
  * Do not compile fop during hyph jar generation
  * Update java source/target version to 1.7
  * Update fop man page
  * Remove jar files from upstream source tarball
  * Bump Std-Vers to 4.2.1 no changes needed

  [ Emmanuel Bourg ]
  * maven: Specify which binary package to build
  * Package libbatik-java does not contain Maven dependency
  * Maven: Skip compilation of source file if artifactId missing
  * maven-war-plugin: war in pom

 -- Mathieu Malaterre <malat@debian.org>  Tue, 20 Nov 2018 18:09:05 +0000

fop (1:2.1-8) unstable; urgency=medium

  * Team upload.
  * Use compat level 9 at least.
  * Fix FTBFS with Java 9. (Closes: #893150)
  * Move the package to salsa.debian.org.

 -- Markus Koschany <apo@debian.org>  Mon, 30 Apr 2018 22:42:07 +0200

fop (1:2.1-7) unstable; urgency=medium

  * Team upload.
  * Fixed the build failure caused by the latest qdox update (Closes: #871106)

 -- Emmanuel Bourg <ebourg@apache.org>  Sun, 06 Aug 2017 15:36:06 +0200

fop (1:2.1-6) unstable; urgency=high

  * Team upload.
  * Fixed CVE-2017-5661: Information disclosure vulnerability (Closes: #860567)

 -- Emmanuel Bourg <ebourg@apache.org>  Wed, 24 May 2017 15:53:03 +0200

fop (1:2.1-5) unstable; urgency=medium

  * Fix FTBFS now that comments are parsed by javadoc. Closes: #852910

fop-core/src/main/java/org/apache/fop/tools/anttasks/Fop.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id: Fop.java 1805173 2017-08-16 10:50:04Z ssteiner $ */

package org.apache.fop.tools.anttasks;

// Ant
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.URI;
import java.util.List;

import org.xml.sax.SAXException;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.impl.SimpleLog;
import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.DirectoryScanner;
import org.apache.tools.ant.Project;
import org.apache.tools.ant.Task;
import org.apache.tools.ant.types.FileSet;
import org.apache.tools.ant.util.GlobPatternMapper;

import org.apache.fop.apps.FOPException;
import org.apache.fop.apps.FOUserAgent;
import org.apache.fop.apps.FopConfParser;
import org.apache.fop.apps.FopFactory;
import org.apache.fop.apps.FopFactoryBuilder;
import org.apache.fop.apps.MimeConstants;
import org.apache.fop.cli.InputHandler;

/**
 * Wrapper for FOP which allows it to be accessed from within an Ant task.
 * Accepts the inputs:
 * <ul>
 * <li>fofile - formatting objects file to be transformed</li>
 * <li>iffile - intermediate format input file to be transformed</li>
 * <li>outputFormat - MIME type of the format to generate ex. "application/pdf"</li>
 * <li>inputFormat - Input file format (like "fo","xml","if")</li>
 * <li>outfile - output filename</li>
 * <li>baseDir - directory to work from</li>
 * <li>relativebase - (true | false) control whether to use each FO's
 *      directory as base directory. false uses the baseDir parameter.</li>
 * <li>userconfig - file with user configuration (same as the "-c" command
 *      line option)</li>
 * <li>messagelevel - (error | warn | info | verbose | debug) level to output
 *      non-error messages</li>
 * <li>logFiles - Controls whether the names of the files that are processed
 *      are logged or not</li>
 * </ul>
 */
public class Fop extends Task {

    private File foFile;
    private File xmlFile;
    private File xsltFile;
    private String xsltParams;
    private List<FileSet> filesets = new java.util.ArrayList<FileSet>();
    private File outFile;
    private File outDir;
    private String format; //MIME type
    private File baseDir;
    private File userConfig;
    private int messageType = Project.MSG_VERBOSE;
    private boolean logFiles = true;
    private boolean force;
    private boolean relativebase;
    private boolean throwExceptions = true;

    /**
     * Sets the filename for the userconfig.xml.
     * @param userConfig Configuration to use
     */
    public void setUserconfig(File userConfig) {
        this.userConfig = userConfig;
    }

    /**
     * Returns the file for the userconfig.xml.
     * @return the userconfig.xml file
     */
    public File getUserconfig() {
        return this.userConfig;
    }

    /**
     * Sets the input XSL-FO file.
     * @param foFile input XSL-FO file
     */
    public void setFofile(File foFile) {
        this.foFile = foFile;
    }

    /**
     * Gets the input XSL-FO file.
     * @return input XSL-FO file
     */
    public File getFofile() {
        return foFile;
    }

    /**
     * Gets the input XML file.
     * @return the input XML file.
     */
    public File getXmlFile() {
        return xmlFile;
    }

    /**
     * Sets the input XML file.
     * @param xmlFile the input XML file.
     */
    public void setXmlFile(File xmlFile) {
        this.xmlFile = xmlFile;
    }

    /**
     * Gets the input XSLT file.
     * @return the input XSLT file.
     */
    public File getXsltFile() {
        return xsltFile;
    }

    /**
     * Sets the input XSLT file.
     * @param xsltFile the input XSLT file.
     */
    public void setXsltFile(File xsltFile) {
        this.xsltFile = xsltFile;
    }

    /**
     * Gets the XSLT parameters
     * @return the XSLT parameters
     */
    public String getXsltParams() {
        return xsltParams;
    }

    /**
     * Sets the XSLT parameters
     * @param xsltParams the XSLT parameters
     */
    public void setXsltParams(String xsltParams) {
        this.xsltParams = xsltParams;
    }

    /**
     * Adds a set of XSL-FO files (nested fileset attribute).
     * @param set a fileset
     */
    public void addFileset(FileSet set) {
        filesets.add(set);
    }

    /**
     * Returns the current list of filesets.
     * @return the filesets
     */
    public List getFilesets() {
        return this.filesets;
    }

    /**
     * Set whether to include files (external-graphics, instream-foreign-object)
     * from a path relative to the .fo file (true) or the working directory (false, default)
     * only useful for filesets
     *
     * @param relbase true if paths are relative to file.
     */
    public void setRelativebase(boolean relbase) {
        this.relativebase = relbase;
    }

    /**
     * Gets the relative base attribute
     * @return the relative base attribute
     */
    public boolean getRelativebase() {
        return relativebase;
    }

    /**
     * Set whether to check dependencies, or to always generate;
     * optional, default is false.
     *
     * @param force true if always generate.
     */
    public void setForce(boolean force) {
        this.force = force;
    }

    /**
     * Gets the force attribute
     * @return the force attribute
     */
    public boolean getForce() {
        return force;
    }

    /**
     * Sets the output file.
     * @param outFile File to output to
     */
    public void setOutfile(File outFile) {
        this.outFile = outFile;
    }

    /**
     * Gets the output file.
     * @return the output file
     */
    public File getOutfile() {
        return this.outFile;
    }

    /**
     * Sets the output directory.
     * @param outDir Directory to output to
     */
    public void setOutdir(File outDir) {
        this.outDir = outDir;
    }

    /**
     * Gets the output directory.
     * @return the output directory
     */
    public File getOutdir() {
        return this.outDir;
    }

    /**
     * Sets output format (MIME type).
     * @param format the output format
     */
    public void setFormat(String format) {
        this.format = format;
    }

    /**
     * Gets the output format (MIME type).
     * @return the output format
     */
    public String getFormat() {
        return this.format;
    }

    /**
     * Set whether exceptions are thrown.
     * default is false.
     * @param throwExceptions true if exceptions should be thrown
     */
    public void setThrowexceptions(boolean throwExceptions) {
        this.throwExceptions = throwExceptions;
    }

    /**
     * Gets the throw exceptions attribute
     * @return the throw exceptions attribute
     */
    public boolean getThrowexceptions() {
        return this.throwExceptions;
    }

    /**
     * Sets the message level to be used while processing.
     * @param messageLevel (error | warn| info | verbose | debug)
     */
    public void setMessagelevel(String messageLevel) {
        if (messageLevel.equalsIgnoreCase("info")) {
            messageType = Project.MSG_INFO;
        } else if (messageLevel.equalsIgnoreCase("verbose")) {
            messageType = Project.MSG_VERBOSE;
        } else if (messageLevel.equalsIgnoreCase("debug")) {
            messageType = Project.MSG_DEBUG;
        } else if (messageLevel.equalsIgnoreCase("err")
                 || messageLevel.equalsIgnoreCase("error")) {
            messageType = Project.MSG_ERR;
        } else if (messageLevel.equalsIgnoreCase("warn")) {
            messageType = Project.MSG_WARN;
        } else {
            log("messagelevel set to unknown value \"" + messageLevel
                + "\"", Project.MSG_ERR);
            throw new BuildException("unknown messagelevel");
        }
    }

    /**
     * Returns the message type corresponding to Project.MSG_*
     * representing the current message level.
     * @return message type
     * @see org.apache.tools.ant.Project
     */
    public int getMessageType() {
        return messageType;
    }

    /**
     * Sets the base directory for single FO file (non-fileset) usage
     * @param baseDir File to use as a working directory
     */
    public void setBasedir(File baseDir) {
        this.baseDir = baseDir;
    }

    /**
     * Gets the base directory.
     * @return the base directory
     */
    public File getBasedir() {
        return (baseDir != null) ? baseDir : getProject().resolveFile(".");
    }

    /**
     * Controls whether the filenames of the files that are processed are logged
     * or not.
     * @param logFiles True if the feature should be enabled
     */
    public void setLogFiles(boolean logFiles) {
        this.logFiles = logFiles;
    }

    /**
     * Returns True if the filename of each file processed should be logged.
     * @return True if the filenames should be logged.
     */
    public boolean getLogFiles() {
        return this.logFiles;
    }

    /**
     * {@inheritDoc}
     */
    public void execute() throws BuildException {
        int logLevel = SimpleLog.LOG_LEVEL_INFO;
        switch (getMessageType()) {
            case Project.MSG_DEBUG  : logLevel = SimpleLog.LOG_LEVEL_DEBUG; break;
            case Project.MSG_INFO   : logLevel = SimpleLog.LOG_LEVEL_INFO; break;
            case Project.MSG_WARN   : logLevel = SimpleLog.LOG_LEVEL_WARN; break;
            case Project.MSG_ERR    : logLevel = SimpleLog.LOG_LEVEL_ERROR; break;
            case Project.MSG_VERBOSE: logLevel = SimpleLog.LOG_LEVEL_DEBUG; break;
            default: logLevel = SimpleLog.LOG_LEVEL_INFO;
        }
        SimpleLog logger = new SimpleLog("FOP/Anttask");
        logger.setLevel(logLevel);
        try {
            FOPTaskStarter starter = new FOPTaskStarter(this);
            starter.setLogger(logger);
            starter.run();
        } catch (FOPException ex) {
            throw new BuildException(ex);
        } catch (IOException ioe) {
            throw new BuildException(ioe);
        } catch (SAXException saxex) {
            throw new BuildException(saxex);
        }

    }

}

class FOPTaskStarter {

    // configure fopFactory as desired
    private final FopFactory fopFactory;

    private Fop task;
    private URI baseUri;

    /**
     * logging instance
     */
    protected Log logger;


    /**
     * Sets the Commons-Logging instance for this class
     * @param logger The Commons-Logging instance
     */
    public void setLogger(Log logger) {
        this.logger = logger;
    }

    /**
     * Returns the Commons-Logging instance for this class
     * @return  The Commons-Logging instance
     */
    protected Log getLogger() {
        return logger;
    }

    FOPTaskStarter(Fop task) throws SAXException, IOException {
        this.task = task;
        //Set base directory
        if (task.getBasedir() != null) {
            this.baseUri = task.getBasedir().toURI();
        } else {
            if (task.getFofile() != null) {
                this.baseUri = task.getFofile().getParentFile().toURI();
            }
        }
        if (task.getUserconfig() != null) {
            FopFactoryBuilder confBuilder = new FopConfParser(
                    task.getUserconfig()).getFopFactoryBuilder();
            confBuilder.setBaseURI(baseUri);
            fopFactory = confBuilder.build();
        } else {
            fopFactory = FopFactory.newInstance(baseUri);
        }
    }

    private static final String[][] SHORT_NAMES = {
        {"pdf",  MimeConstants.MIME_PDF},
        {"ps",   MimeConstants.MIME_POSTSCRIPT},
        {"mif",  MimeConstants.MIME_MIF},
        {"rtf",  MimeConstants.MIME_RTF},
        {"pcl",  MimeConstants.MIME_PCL},
        {"txt",  MimeConstants.MIME_PLAIN_TEXT},
        {"at",   MimeConstants.MIME_FOP_AREA_TREE},
        {"xml",  MimeConstants.MIME_FOP_AREA_TREE},
        {"tiff", MimeConstants.MIME_TIFF},
        {"tif",  MimeConstants.MIME_TIFF},
        {"png",  MimeConstants.MIME_PNG},
        {"afp",  MimeConstants.MIME_AFP}
    };

    private String normalizeOutputFormat(String format) {
        if (format == null) {
            return MimeConstants.MIME_PDF;
        }
        for (String[] shortName : SHORT_NAMES) {
            if (shortName[0].equals(format)) {
                return shortName[1];
            }
        }
        return format; //no change
    }

    private static final String[][] EXTENSIONS = {
        {MimeConstants.MIME_FOP_AREA_TREE,   ".at.xml"},
        {MimeConstants.MIME_FOP_AWT_PREVIEW, null},
        {MimeConstants.MIME_FOP_PRINT,       null},
        {MimeConstants.MIME_PDF,             ".pdf"},
        {MimeConstants.MIME_POSTSCRIPT,      ".ps"},
        {MimeConstants.MIME_PCL,             ".pcl"},
        {MimeConstants.MIME_PCL_ALT,         ".pcl"},
        {MimeConstants.MIME_PLAIN_TEXT,      ".txt"},
        {MimeConstants.MIME_RTF,             ".rtf"},
        {MimeConstants.MIME_RTF_ALT1,        ".rtf"},
        {MimeConstants.MIME_RTF_ALT2,        ".rtf"},
        {MimeConstants.MIME_MIF,             ".mif"},
        {MimeConstants.MIME_SVG,             ".svg"},
        {MimeConstants.MIME_PNG,             ".png"},
        {MimeConstants.MIME_JPEG,            ".jpg"},
        {MimeConstants.MIME_TIFF,            ".tif"},
        {MimeConstants.MIME_AFP,             ".afp"},
        {MimeConstants.MIME_AFP_ALT,         ".afp"},
        {MimeConstants.MIME_XSL_FO,          ".fo"}
    };

    private String determineExtension(String outputFormat) {
        for (String[] extension : EXTENSIONS) {
            if (extension[0].equals(outputFormat)) {
                String ext = extension[1];
                if (ext == null) {
                    throw new RuntimeException("Output format '"
                            + outputFormat + "' does not produce a file.");
                } else {
                    return ext;
                }
            }
        }
        return ".unk"; //unknown
    }

    private File replaceExtension(File file, String expectedExt,
                                  String newExt) {
        String name = file.getName();
        if (name.toLowerCase().endsWith(expectedExt)) {
            name = name.substring(0, name.length() - expectedExt.length());
        }
        name = name.concat(newExt);
        return new File(file.getParentFile(), name);
    }

    public void run() throws FOPException {
        task.log("Using base URI: " + baseUri, Project.MSG_DEBUG);

        String outputFormat = normalizeOutputFormat(task.getFormat());
        String newExtension = determineExtension(outputFormat);

        // actioncount = # of fofiles actually processed through FOP
        int actioncount = 0;
        // skippedcount = # of fofiles which haven't changed (force = "false")
        int skippedcount = 0;

        // deal with single source file
        if (task.getFofile() != null) {
            if (task.getFofile().exists()) {
                File outf = task.getOutfile();
                if (outf == null) {
                    throw new BuildException("outfile is required when fofile is used");
                }
                if (task.getOutdir() != null) {
                    outf = new File(task.getOutdir(), outf.getName());
                }
                // Render if "force" flag is set OR
                // OR output file doesn't exist OR
                // output file is older than input file
                if (task.getForce() || !outf.exists()
                    || (task.getFofile().lastModified() > outf.lastModified())) {
                    render(task.getFofile(), outf, outputFormat);
                    actioncount++;
                } else if (outf.exists()
                        && (task.getFofile().lastModified() <= outf.lastModified())) {
                    skippedcount++;
                }
            }
        } else if (task.getXmlFile() != null && task.getXsltFile() != null) {
            if (task.getXmlFile().exists() && task.getXsltFile().exists()) {
                File outf = task.getOutfile();
                if (outf == null) {
                    throw new BuildException("outfile is required when fofile is used");
                }
                if (task.getOutdir() != null) {
                    outf = new File(task.getOutdir(), outf.getName());
                }
                // Render if "force" flag is set OR
                // OR output file doesn't exist OR
                // output file is older than input file
                if (task.getForce() || !outf.exists()
                        || (task.getXmlFile().lastModified() > outf.lastModified()
                        || task.getXsltFile().lastModified() > outf.lastModified())) {
                    render(task.getXmlFile(), task.getXsltFile(), outf, outputFormat);
                    actioncount++;
                } else if (outf.exists()
                        && (task.getXmlFile().lastModified() <= outf.lastModified()
                            || task.getXsltFile().lastModified() <= outf.lastModified())) {
                    skippedcount++;
                }
            }
        }

        GlobPatternMapper mapper = new GlobPatternMapper();

        String inputExtension = ".fo";
        File xsltFile = task.getXsltFile();
        if (xsltFile != null) {
            inputExtension = ".xml";
        }
        mapper.setFrom("*" + inputExtension);
        mapper.setTo("*" + newExtension);

        // deal with the filesets
        for (int i = 0; i < task.getFilesets().size(); i++) {
            FileSet fs = (FileSet) task.getFilesets().get(i);
            DirectoryScanner ds = fs.getDirectoryScanner(task.getProject());
            String[] files = ds.getIncludedFiles();

            for (String file : files) {
                File f = new File(fs.getDir(task.getProject()), file);

                File outf = null;
                if (task.getOutdir() != null && file.endsWith(inputExtension)) {
                    String[] sa = mapper.mapFileName(file);
                    outf = new File(task.getOutdir(), sa[0]);
                } else {
                    outf = replaceExtension(f, inputExtension, newExtension);
                    if (task.getOutdir() != null) {
                        outf = new File(task.getOutdir(), outf.getName());
                    }
                }
                File dir = outf.getParentFile();
                if (!dir.exists()) {
                    dir.mkdirs();
                }
                try {
                    if (task.getRelativebase()) {
                        this.baseUri = f.getParentFile().toURI();
                    }
                    if (this.baseUri == null) {
                        this.baseUri = fs.getDir(task.getProject()).toURI();
                    }

                } catch (Exception e) {
                    task.log("Error setting base URL", Project.MSG_DEBUG);
                }

                // Render if "force" flag is set OR
                // OR output file doesn't exist OR
                // output file is older than input file
                if (task.getForce() || !outf.exists()
                        || (f.lastModified() > outf.lastModified())) {
                    if (xsltFile != null) {
                        render(f, xsltFile, outf, outputFormat);
                    } else {
                        render(f, outf, outputFormat);
                    }
                    actioncount++;
                } else if (outf.exists() && (f.lastModified() <= outf.lastModified())) {
                    skippedcount++;
                }
            }
        }

        if (actioncount + skippedcount == 0) {
            task.log("No files processed. No files were selected by the filesets "
                + "and no fofile was set." , Project.MSG_WARN);
        } else if (skippedcount > 0) {
            task.log(skippedcount + " xslfo file(s) skipped (no change found"
                + " since last generation; set force=\"true\" to override)."
                , Project.MSG_INFO);
        }
    }

    private void renderInputHandler(InputHandler inputHandler, File outFile, String outputFormat)
            throws Exception {
        OutputStream out = null;
        try {
            out = new java.io.FileOutputStream(outFile);
            out = new BufferedOutputStream(out);
        } catch (Exception ex) {
            throw new BuildException("Failed to open " + outFile, ex);
        }
        boolean success = false;
        try {
            FOUserAgent userAgent = fopFactory.newFOUserAgent();
            inputHandler.renderTo(userAgent, outputFormat, out);
            success = true;
        } catch (Exception ex) {
            if (task.getThrowexceptions()) {
                throw new BuildException(ex);
            }
            throw ex;
        } finally {
            try {
                out.close();
            } catch (IOException ioe) {
                logger.error("Error closing output file", ioe);
            }
            if (!success) {
                outFile.delete();
            }
        }
    }

    private void render(File foFile, File outFile,
                        String outputFormat) throws FOPException {
        InputHandler inputHandler = new InputHandler(foFile);
        try {
            renderInputHandler(inputHandler, outFile, outputFormat);
        } catch (Exception ex) {
            logger.error("Error rendering fo file: " + foFile, ex);
        }
        if (task.getLogFiles()) {
            task.log(foFile + " -> " + outFile, Project.MSG_INFO);
        }
    }

    private void render(File xmlFile, File xsltFile, File outFile, String outputFormat) {
        //TODO: implement support for XSLT params
        InputHandler inputHandler = new InputHandler(xmlFile, xsltFile, null);
        try {
            renderInputHandler(inputHandler, outFile, outputFormat);
        } catch (Exception ex) {
            logger.error("Error rendering xml/xslt files: "
                         + xmlFile + ", " + xsltFile, ex);
        }
        if (task.getLogFiles()) {
            task.log("xml: " + xmlFile + ", xslt: "
                     + xsltFile + " -> " + outFile, Project.MSG_INFO);
        }
    }
}

